exercises/practice/dot-dsl/src/main/java/DotDsl.java

```diff
@@ -0,0 +1,65 @@
+import java.util.Collection;
+import java.util.Map;
+
+class DotDsl {
+
+    static class Graph {
+        public Graph() {
+        }
+
+        public Graph(Map<String, String> attributes) {
+        }
+
+        public Collection<Node> getNodes() {
+            throw new UnsupportedOperationException();
+        }
+
+        public Collection<Edge> getEdges() {
+            throw new UnsupportedOperationException();
+        }
+
+        public Graph node(String name) {
+            throw new UnsupportedOperationException();
+        }
+
+        public Graph node(String name, Map<String, String> attributes) {
+            throw new UnsupportedOperationException();
+        }
+
+        public Graph edge(String start, String end) {
+            throw new UnsupportedOperationException();
+        }
+
+        public Graph edge(String start, String end, Map<String, String> attributes) {
+            throw new UnsupportedOperationException();
+        }
+
+        public Map<String, String> getAttributes() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    static class Node {
```

Comment:
I was thinking as this practice exercise is already one of the hard ones, maybe we can simplify the task of the user only having them to implement `Graph` and `Element` and we give the `Edge` and `Node` already implemented.

We can use the [tree-building](https://github.com/exercism/java/tree/85e331f85bbace9b4227bf7f6e33f5ee12b5fd1b/exercises/practice/tree-building) practice exercise as an example that gives the `Record` and `TreeNode` implementations.

Reply:
I thought about that while implementing the exercise. I chose to leave it out because it is still technically possible to implement a solution _without_ the `Element` class (by duplicating the methods in the `Graph`, `Node` and `Edge` classes). I thought this could be discussion point between a mentor and student or, possibly, for the analyzer. 

Let me know if you'd still prefer me to add `Element`. I'm thinking I'd give them the skeleton rather the full `Element` implementation as the student's task in this one is to implement the graph classes. I think this is different to `tree-building`, which is a a _refactoring_ exercise (so it actually starts with a working solution and the student's task is to try make it better). 

I probably should make `inheritance` a prerequisite though. That way the student will have been introduced to the concept before this exercise.

Reply:
Maybe [dominoes](https://github.com/exercism/java/tree/main/exercises/practice/dominoes) is a better example of what I was trying to say. They give the `Domino` class implemented and the `Dominoes` to be implemented.

For example if we give `Edge` and `Node` already implemented, the user has multiple options to solve the exercise? or only the one that they use the `Element` class?

Reply:
I have been thinking about this one the last couple of days. I think providing them with the `Edge` and `Node` implementations allows us to make the assertions easier to follow because we can implement their `equals` and `hashCode`. We also wouldn't need `Element` anymore.

Anyway, I thought I try this out and have just pushed the commits for this in. The commit also:
- Reduces the difficulty as I think the exercise has become easier.
- Changes the "inheritance" prerequisite back to "classes" as inheritance is no longer applicable. 

Let me know what you think.

Reply:
Love this one!